### ISSUE TITLE
Promises randomly cease working ~once a month?

### ISSUE THREAD

**Issue by kentonv** [stefanpenner/es6-promise#159]
I'm filing this on the off chance that someone looks and says: "Oh yeah, I know why that would happen -- it's because of X!" More likely, no one will have any idea, and I'll keep debugging, slowly.

I have a node-0.10.x-based server in which I'm using [this version of es6-promise](https://es6-promises.s3.amazonaws.com/es6-promise-2.0.1.min.js) (apparently).

Approximately once per machine-month, a problem happens where the promise library spontaneously decides not to call continuations anymore. That is, I have code like this:

    var promise = new Promise(function (resolve, reject) {
      doSomeAsyncThing(function (x) {
        console.log(1);
        resolve(x);
      });
    }).then(function (response) {
      console.log(2);
    });

Normally when it prints `1` it also prints `2`, but after this rare problem occurs, only `1` is printed, without a corresponding `2`. This seems to affect all promises across the server process. Once it has started happening, the only remedy I've found is to kill the process. I've only shown the "resolve" path above but "reject" shows the same behavior -- the error callback never executes.

Unfortunately, I have not been able to determine how to reproduce the problem. Anecdotally, the problem seems to occur more frequently when load is high, but forced load tests do not reproduce it. So, I have been very slowly debugging by adding logging statements and waiting for the problem to happen again in production. It has taken about a year to narrow the problem down to es6-promise (initially I was convinced the problem had to be somewhere in libuv, as initially I thought that the entire event loop had stopped, not just Promise-based events).

I realize that 2.0.1 is old. Perhaps upgrading to the latest version would fix the problem. However, I feel it is important to actually determine the root cause, and I don't see any changes in the commit log that look relevant in any case.

From reading the code a bit, it seems to me most likely that `asap()` is somehow ceasing to execute new events, since it appears to be the only place in the library that has global mutable state. Indeed, if `len` could somehow become non-zero without a flush being scheduled, or if a flush were missed somehow, that would explain the problem. However, I can't see any bugs in the code which could cause this.

Another possibility is that the library is somehow being "accidentally monkey-patched" by other code in the process, causing it to be broken, but that seems relatively hard to do by accident, and why would it be so rare?

I'll continue debugging and update this if I find the problem, but if anyone happens to have any insights, please share.

**Comment by stefanpenner** [comment]
It is most likely an issue with libuv/node or something OS specifc (as libuv ultimately delegates to w/e the OS provides) as es6-promise merely delegates to `process.nextTick`

**Comment by kentonv** [comment]
@stefanpenner Well, maybe. If `nextTick()` were entirely non-functional then I would expect the server not to function at all. As it is, only the parts that depend on es6-promise seem to stop working.

I suppose, though, that if a bug in libuv/node very occasionally causes `nextTick()` to fail to call its callback, and one of these dropped `nextTick()` calls happened to be from `asap()`, then `asap()` would be rendered permanently broken in that process, due to its implementation. So possibly that's what is happening... though `nextTick()` dropping events seems like a pretty severe node bug even if it's rare.

**Comment by stefanpenner** [comment]
@kentonv yes 1 dropped nextTick could put `asap` into a bad state, ultimately any number of dropped `nextTick` will put a system in an undefined state :(.

In theory it is possible for us to add additional machinery that can perform a health-check + recovery, but I am nervous to do so without confirming the dropped nextTick is actually the problem.

Are their any exceptions in your failure cases, or is it unknown?


--- 

Some thoughts:

* [Resetting length at the end of the loop](https://github.com/jakearchibald/es6-promise/blob/master/lib/es6-promise/asap.js#L94) may be risky, although this would only be an issue if the internal promise could itself also throw an exception... which should result in a crashed process as it would be thrown on a fresh frame.

**Comment by stefanpenner** [comment]
It's also worth noting, other performant promise implementations would also be affected, as they typically take a similar approach to batching `nextTick`, if they don't mysterously dropped `nextTick` in a non batched environment, would still cause potentially serious grief but wouldn't cause all future tasks to also fail.

**Comment by kentonv** [comment]
To be clear, I'm definitely **not** advocating for es6-promise to work around bugs in Node. If it is dropping nextTick()s, that's a serious problem that needs to be fixed at the source.

But for the moment, my priority is figuring out what the root cause is, then we can decide how to fix it. :)

**Comment by kentonv** [comment]
BTW I was also suspicious initially of what might happen if a callback threw an exception during `flush()`, causing the `len = 0` line to be skipped. However, I'm _pretty_ sure that an exception during a `nextTick()` callback would in fact kill the process, or at least print a message. I have not observed any suspicious exceptions printed in the logs, and definitely haven't observed process deaths.

**Comment by stefanpenner** [comment]
I'm not sure what we can do here :( Any further details?

**Comment by kentonv** [comment]
The bug has been mysteriously quiet of late. I see no occurrences over all our machines in the last month, which is as far back as our logs go. Maybe it was fixed, although I'm not aware of anything I changed. I may never know. In a couple months, Meteor 1.4 will come out and finally allow us to use Node 4.x which means we won't be using this polyfill anymore, which almost certainly means the problem will be gone by then if it isn't already. I don't actually know what to try next as far as debugging so maybe it's time to give up?
